### PR TITLE
Sprint 2 review

### PR DESCRIPTION
**Pull request by randyhart**
Sprint 2
Stories #126, 128, 71

This was filed as [this pull request](https://github.com/ustaxcourt/ef-cms/pull/7) originally, but @waldoj erroneously merged it.

### CODE REVIEW

**Comment by waldoj** [comment]
`efcms-service`'s `npm start` doesn’t work without running `npm run install:dynamodb`, but that’s not in the `README.md`. Please add that instruction.

**Comment by waldoj** [comment]
Immediately upon logging on, the site complains `User not found`, even though the login process worked fine:

not_found

Steps to reproduce:

1. Go to `/log-in`.
2. Enter `taxpayer`.
3. Observe the `User not found` error.

**Comment by waldoj** [comment]
I'd like to see either `README.md` or `SETUP.md` include a bit of information about how to stand up the site locally, even if it's as simple as advising people to follow the instructions in the `README.md` files found in `document-service/` and `web-client/`, to run both the backend and the front end in a pair of terminal windows.

**Comment by waldoj** [comment]
Accessibility concern: the ARIA label for the `Pay now` button reads `pay.gov us tax court filing fees`, which is narrated by VoiceOver as "pay dot gov us," instead of "pay.gov you ess." Text readers are not consistent in this kind of thing (no doubt others would read it as "you ess"), but making this explicitly `u.s.` instead of `us` should guarantee that this is read correctly.

**Comment by waldoj** [comment]
Accessibility concern: The layout of the Case Activity Record is sensible for sighted users, but not for blind users. The `Save Updates` button is in the top right, so after selecting the `Validate` checkboxes, it is necessary to navigate back up to save the updates.

screenshot

If I might suggest a solution, the tabindex can be used to ensure that `Save updates` follows the final `Validate` checkbox.

File: efcms-service/deploy-sandbox.sh [hunk 1]
```diff
@@ -47,11 +58,11 @@ function configure_custom_logging() {
...
         --rest-api-id "${REST_API_ID}" \
         --stage-name "${ENVIRONMENT}" \
-        --region us-east-1 \
+        --region "${REGION}" \
         --patch-operations op=replace,path=/*/*/logging/dataTrace,value=true
 }
 
 check_env_vars_exist
-prepare_serverless
+# prepare_serverless
```
[the remark this answers is not in the document]
Comment (codyseibert) [reply efcms-service/deploy-sandbox.sh]: being addressed

File: efcms-service/src/cases/getCase.test.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        headers: { 'Authorization': 'Bearer userId' }
+      }
+    ].forEach(function (documentBody) {
+      it('should return a case on a GET', function () {
+        return lambdaTester(getCase.get)
+          .event(documentBody)
+          .expectResolve(result => {
+            const data = JSON.parse(result.body);
+            expect(data).to.equal('Case 123 was not found.');
```
[the remark this answers is not in the document]
Comment (codyseibert) [reply efcms-service/src/cases/getCase.test.js]: being addressed

File: efcms-service/src/cases/getCases.test.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        httpMethod: 'GET',
+        queryStringParameters: {'status': 'new'},
+        headers: { 'Authorization': '' }
+      }
+    ].forEach(function (documentBody) {
+      it('should return a 404 if no user is present', function () {
+        return lambdaTester(getCases.get)
+          .event(documentBody)
+          .expectResolve(error => {
+            expect(error.statusCode).to.equal('400');
```
[the remark this answers is not in the document]
Comment (codyseibert) [reply efcms-service/src/cases/getCases.test.js]: being addressed

File: efcms-service/src/cases/middleware/caseMiddleware.test.js [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+    after(() => {
+      client.query.restore();
+    })
+
+    it('should get all cases for a user', async () => {
+      const result = await caseMiddleWare.getCases({
+        userId: 'user'
+      });
+      expect(result[0]).to.equal(item);
```
[the remark this answers is not in the document]
Comment (codyseibert) [reply efcms-service/src/cases/middleware/caseMiddleware.test.js]: being addressed

File: efcms-service/src/cases/middleware/caseMiddleware.test.js [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+  it('should get cases by query string status when user is authorized', async () => {
+    const result = await caseMiddleWare.getCasesByStatus({
+      status: 'NEW',
+      userId: 'petitionsclerk'
+    });
+    expect(result[0]).to.equal(item);
+  });
+
+  it('should return an error if the user is authorized', async () => {
```
[the remark this answers is not in the document]
Comment (codyseibert) [reply efcms-service/src/cases/middleware/caseMiddleware.test.js]: being addressed

File: efcms-service/src/middleware/authorizationClientService.test.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+  it('should authorize a user who is owner always', () => {
+    expect(isAuthorized('someUser', 'unknown action', 'someUser')).to.be.true;
+  });
+
+  it('should authorize a petitionsclerk for getCase', () => {
+    expect(isAuthorized('petitionsclerk', GET_CASE)).to.be.true;
+  });
+
+  it('should set the message', () => {
```
[the remark this answers is not in the document]
Comment (codyseibert) [reply efcms-service/src/middleware/authorizationClientService.test.js]: being addressed

File: efcms-service/src/middleware/errors.test.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+});
+
+describe('UnprocessableEntitytError', () => {
+  let error;
+
+  beforeEach(() => {
+    error = new UnprocessableEntityError();
+  });
+
+  it('should set a status code of 404', () => {
```
[the remark this answers is not in the document]
Comment (codyseibert) [reply efcms-service/src/middleware/errors.test.js]: being addressed

File: efcms-service/src/middleware/errors.test.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    expect(error.statusCode).to.equal(404);
+  });
+
+  it('should set the message', () => {
+    expect(error.message).to.equal("some error");
+  });
+});
+
+describe('UnprocessableEntitytError', () => {
```
[the remark this answers is not in the document]
Comment (codyseibert) [reply efcms-service/src/middleware/errors.test.js]: being addressed

**Comment by kkoskelin** [comment]
> Accessibility concern: Selecting (not clicking) the `Upload` button causes the entire contents of the page to be read again. Steps to reproduce:
> 
> 1. Enable VoiceOver.
> 2. Log in and go to the upload form.
> 3. Tab down to the `Upload` button.

> 4. Observe that, after VoiceOver speaks "Upload," it then reads all of the contents of the page again.

This behavior was replicated when using MacOS Voiceover (henceforth VO) + Chrome 70 when using the TAB key to focus the "Upload" button.  If one is using VO + Chrome and using the VO keyboard navigation shortcuts to navigate to the "Upload" button, it does not occur.

At this time the behavior seems to be related to the use of `role="list"` and `role="listitem"` in the form. If `role="list"` is removed from the parent DIV element, Chrome no longer exhibits this behavior.

The presence of roles makes the form more accessible and it has been decided to keep them.  The behavior describe is *not* seen when using VO + Safari (which is the preferred method of testing VO).

cc: @waldoj

**Comment by waldoj** [comment]
All issues have been resolved!

**Pull request merged**